Refactor lab extension

jasongrout opened the pull request:
Make it so that we export the token/interface from the main package, and the CSS is only imported in the plugin. This ensures that downstream plugins don’t include the CSS (which they shouldn’t) and only import from the main package instead of a module deep in our package (which they should).

jasongrout commented:
Don't merge this - it's still a work in progress as I work on downstream extensions.

jasongrout commented:
Okay, I think this is ready. I also changed the class names from things like `IIPyWidget...` to `INBWidget` to match the package name `@jupyterlab/nbwidgets`.

jasongrout commented:
@gnestor, @SylvainCorlay - ready for review.

gnestor commented:
The changes look fine to me. Is there a way to test this?

jasongrout commented:
I've been testing against the bqplot widget extension. We don't have unit tests for this yet.

jasongrout commented:
Well, one test is if it works with JupyterLab still.

gnestor commented:
Right, I just tested bqplot with jupyterlab and I'm getting the `Widget Javascript not detected.  It may not be installed properly. Did you enable the widgetsnbextension? If not, then run "jupyter nbextension enable --py --sys-prefix widgetsnbextension"` error (same with ipywidgets)...

jasongrout commented:
> Did you enable the widgetsnbextension?

Yeah, we should change that message now :).

Check to make sure your widgets lab extension is enabled and is being loaded.

gnestor commented:
Oh I didn't realize that `jupyter labextension` was a thing. I followed [these instructions](https://github.com/ipython/ipywidgets/tree/master/jupyterlab_widgets#source-build) to install. What's the command to install the labextension for ipywidgets? Bqplot?

jasongrout commented:
> What's the command to install the labextension for ipywidgets?

Updated instructions: https://github.com/jasongrout/ipywidgets/blob/59ac6665f2ca7ddcad3c79e094b2822459fe15f9/jupyterlab_widgets/README.md#source-build

Apparently the docs are broken right now too. I'll work on them...

gnestor commented:
Ok, very helpful! However, I'm seeing this error when running `pip install -e .` (after running `npm install` and `npm run build`):

```
rebuilding js and css failed
    missing files: ['/Users/grant/Sites/jupyter/ipywidgets/jupyterlab_widgets/jupyterlab_widgets/static/jupyterlab_widgets.bundle.js']
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/Users/grant/Sites/jupyter/ipywidgets/jupyterlab_widgets/setup.py", line 229, in <module>
        setup(**setup_args)
      File "/Users/grant/anaconda/lib/python3.5/distutils/core.py", line 148, in setup
        dist.run_commands()
      File "/Users/grant/anaconda/lib/python3.5/distutils/dist.py", line 955, in run_commands
        self.run_command(cmd)
      File "/Users/grant/anaconda/lib/python3.5/distutils/dist.py", line 974, in run_command
        cmd_obj.run()
      File "/Users/grant/Sites/jupyter/ipywidgets/jupyterlab_widgets/setup.py", line 97, in run
        raise e
      File "/Users/grant/Sites/jupyter/ipywidgets/jupyterlab_widgets/setup.py", line 90, in run
        self.distribution.run_command('jsdeps')
      File "/Users/grant/anaconda/lib/python3.5/distutils/dist.py", line 974, in run_command
        cmd_obj.run()
      File "/Users/grant/Sites/jupyter/ipywidgets/jupyterlab_widgets/setup.py", line 161, in run
        raise ValueError(msg)
    ValueError: Missing file: /Users/grant/Sites/jupyter/ipywidgets/jupyterlab_widgets/jupyterlab_widgets/static/jupyterlab_widgets.bundle.js
```

`./jupyterlab_widgets/jupyterlab_widgets/static` doesn't exist, so `buildExtension.js` is failing. The error I see when running `npm run build:extension`:

```
Module not found: Error: Cannot resolve module 'jquery-ui/themes/base/all.css' in /Users/grant/Sites/jupyter/ipywidgets/jupyterlab_widgets/lib
    rebuilding js and css failed
    missing files: ['/Users/grant/Sites/jupyter/ipywidgets/jupyterlab_widgets/jupyterlab_widgets/static/jupyterlab_widgets.bundle.js']
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/Users/grant/Sites/jupyter/ipywidgets/jupyterlab_widgets/setup.py", line 229, in <module>
        setup(**setup_args)
      File "/Users/grant/anaconda/lib/python3.5/distutils/core.py", line 148, in setup
        dist.run_commands()
      File "/Users/grant/anaconda/lib/python3.5/distutils/dist.py", line 955, in run_commands
        self.run_command(cmd)
      File "/Users/grant/anaconda/lib/python3.5/distutils/dist.py", line 974, in run_command
        cmd_obj.run()
      File "/Users/grant/Sites/jupyter/ipywidgets/jupyterlab_widgets/setup.py", line 97, in run
        raise e
      File "/Users/grant/Sites/jupyter/ipywidgets/jupyterlab_widgets/setup.py", line 90, in run
        self.distribution.run_command('jsdeps')
      File "/Users/grant/anaconda/lib/python3.5/distutils/dist.py", line 974, in run_command
        cmd_obj.run()
      File "/Users/grant/Sites/jupyter/ipywidgets/jupyterlab_widgets/setup.py", line 161, in run
        raise ValueError(msg)
    ValueError: Missing file: /Users/grant/Sites/jupyter/ipywidgets/jupyterlab_widgets/jupyterlab_widgets/static/jupyterlab_widgets.bundle.js
```

jasongrout commented:
What version of jqueryui do you have in node modules? Is the file in question there?

jasongrout commented:
Actually, it looks like you probably didn't update the jupyter-js-widgets dependency to reflect changes in it. I just added an `npm run update` command in `jupyterlab_widgets` and instructions. Can you do:

```
npm run update
npm run clean
npm run build
pip install -e .
```

gnestor commented:
Ah, it looks like it's called `jquery.ui.all.css` in both `jupyterlab_extension` and `widgetsnbextension` (because jupyter-js-widgets@5.2.2 on npm [depends on jquery-ui@~1.10.5](https://github.com/ipython/ipywidgets/blob/5.x/jupyter-js-widgets/package.json#L68)) and `all.css` in `jupyter-js-widgets` (because master [depends on jquery-ui@^1.12.1](https://github.com/ipython/ipywidgets/blob/master/jupyter-js-widgets/package.json#L82)). The CSS files were all renamed with [this commit on jquery-ui](https://github.com/jquery/jquery-ui/commit/21154cfa2e02ef1814a6aff68b14553bdad165cb) so it should be `all.css`. 

I did a `npm install ../jupyter-js-widgets` and that allowed me to build and I can verify that all widgets are working in jupyterlab except Progress and Accordion (CSS styles are missing). Should `jupyterlab_extension` depend on the local `jupyter-js-widgets` (e.g. `"jupyter-js-widgets": "../jupyter-js-widgets"`)? If not, then we should probably update the dev install docs to reflect this or add a step in one of the build scripts.

Progress:
![image](https://cloud.githubusercontent.com/assets/512354/19691486/4bf3763e-9a89-11e6-9861-b372cc452aa8.png)

Accordion:
![image](https://cloud.githubusercontent.com/assets/512354/19691453/31854c28-9a89-11e6-8a6c-713a1a38ef51.png)

jasongrout commented:
> Should jupyterlab_extension depend on the local jupyter-js-widgets (e.g. "jupyter-js-widgets": "../jupyter-js-widgets")? If not, then we should probably update the dev install docs to reflect this or add a step in one of the build scripts.

That's what widgetsnbextension does. I don't like that it adds an extra step to the release process, and what is released is different then what is in the repo.

I'll release a new jupyter-js-widgets and bump the jupyterlab_widgets dependency. That should take care of the issue.

jasongrout commented:
Okay, I released 2.0.3 of jupyter-js-widgets and bumped the dependency to that.

gnestor commented:
That resolved building jupyterlab_extension for me! Are Progress and Accordion widgets rendering correctly for you?

jasongrout commented:
Are the still not rendering well for you? Did you make sure to refresh your browser cache, to pull in the new css?

jasongrout commented:
Both the accordion and tabs render just fine for me in JupyterLab.

jasongrout commented:
However, the progress bar doesn't seem to be working. I'll look at it.

gnestor commented:
It looked like my jupyterlab_extension is still using the older version of jupyter-js-widgets. I followed the "Rebuild" instructions and now Accordion is rendering correctly but now Sliders are not!

```
TypeError: (intermediate value)(intermediate value).slider is not a function 
at IntSliderView.render (http://localhost:8888/labextension/jupyterlab_widgets/jupyterlab_widgets.bundle.js:52961:15) 
at http://localhost:8888/labextension/jupyterlab_widgets/jupyterlab_widgets.bundle.js:25091:46
```

SylvainCorlay commented:
Guys, I have fixed all these issues in my css variables pr. I got back into it today. Should be finished by tomorrow.

jasongrout commented:
@gnestor - delete your node_modules and npm install again. There may be some old packages hanging around.

jasongrout commented:
@gnestor, @SylvainCorlay - I added the progress bar styles, so this will work now too.  Grant, you'll need to update jupyter-js-widgets.

In jupyter-js-widgets:
```
npm run clean
npm run build
```

In jupyterlab_widgets:
```
npm run clean
npm run update
npm run build
```

I also rebased on master, which should take care of the sliders not working.

When this is merged, we'll bump jupyter-js-widgets to 2.0.4, update jupyterlab_widgets to point to 2.0.4, and bump jupyterlab_widgets and release it as well.

SylvainCorlay commented:
Jason, did you see the above message?

SylvainCorlay commented:
Actually that is alright. You are modifying widgets.less which is not going to conflict with my changes.

jasongrout commented:
> @jasongrout you probably missed the above message

I saw it, which is why I CCd you on my comment. Again, this gets things working for now, so this can be merged, and doesn't block on your css rewrite.

gnestor commented:
@jasongrout I tried the following and also `rm -r node_modules && npm install` in both directories and I'm still not seeing Sliders and Progress widgets:

> In jupyter-js-widgets:
```
npm run clean
npm run build
```
>
In jupyterlab_widgets:
```
npm run clean
npm run update
npm run build
```

My branch is up-to-date with 'jasongrout/labrefactor'. I obviously restarted jupyterlab. I'm not sure what I'm missing... ☹️

SylvainCorlay commented:
Discussed in person with @jasongrout

jasongrout commented:
It looks like it was premature to update the jupyterlab dependency - we are fixing that in Sylvain's CSS PR.
